stan/math/laplace/laplace_marginal.hpp

```diff
@@ -0,0 +1,275 @@
+#ifndef STAN_MATH_LAPLACE_LAPLACE_MARGINAL_HPP
+#define STAN_MATH_LAPLACE_LAPLACE_MARGINAL_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/quad_form_diag.hpp>
+#include <stan/math/prim/mat/fun/diag_pre_multiply.hpp>
+#include <stan/math/prim/mat/fun/diag_post_multiply.hpp>
+#include <stan/math/prim/mat/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/mat/fun/sqrt.hpp>
+#include <stan/math/rev/mat/fun/cholesky_decompose.hpp>
+#include <stan/math/laplace/laplace_likelihood.hpp>
+
+
+// Reference for calculations of marginal and its gradients:
+// Rasmussen and Williams,
+// "Gaussian Processes for Machine Learning", first edition,
+// Algorithms 3.1 and 5.1.
+// The MIT Press, 2006.
+// Note: where I didn't conflict with my own notation, I used their notation.
+
+namespace stan {
+namespace math {
+  /**
+   * Function to compute the log marginal density.
+   * Variables needed for the gradient are stored by reference.
+   *
+   * theta_0: initial guess for latent gaussian variable.
+   * phi: global parameter (input in covariance function)
+   * x: fixed data (input in covariance function)
+   * D: structure to compute the conditional log likelihood and its
+   *    derivatives.
+   * K: a functor to compute the covariance function specified by the user.
+   * theta: mode of the target.
+   * W_root: squared root of the negative diagnonal Hessian.
+   * L: Cholesky decomposition of B, the stabilized covariance inverse.
+   * a: component in the Newton step.
+   * tolerance: function tolerance defined in terms of objective function.
+   * max_num_steps: number of Newton iterations after which the algorithm
+   *                stops.
+   */
```

Comment:
Again: I haven't written the docs, because all of this code is COMPLETELY EXPERIMENTAL!!!! I'm still tweaking it, and the docs is the last thing I'll write. Didn't put in throws and returns yet.

Reply:
lol my b. I'd recommend doc'ing as you go so that things are fresher when you write the docs. that and docs can become sort of a big ugh if you leave it as the last thing to do all at once